src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs

```diff
@@ -5437,6 +5462,1076 @@ class CL1<T>
+                //     partial void M2([NotNullWhenFalse] string? s);
+                Diagnostic(ErrorCode.ERR_AttributeRequiresBoolReturn, "NotNullWhenFalse").WithArguments("NotNullWhenFalseAttribute").WithLocation(8, 22),
+                // (12,22): error CS8627: The 'NotNullWhenFalseAttribute' attribute is only applicable on members that return a boolean type.
+                //     partial void M3([NotNullWhenFalse] string? s) => throw null;
+                Diagnostic(ErrorCode.ERR_AttributeRequiresBoolReturn, "NotNullWhenFalse").WithArguments("NotNullWhenFalseAttribute").WithLocation(12, 22),
+                // (11,22): error CS0579: Duplicate 'NotNullWhenFalse' attribute
+                //     partial void M3([NotNullWhenFalse] string? s);
+                Diagnostic(ErrorCode.ERR_DuplicateAttribute, "NotNullWhenFalse").WithArguments("NotNullWhenFalse").WithLocation(11, 22),
+                // (6,22): error CS8627: The 'NotNullWhenFalseAttribute' attribute is only applicable on members that return a boolean type.
+                //     partial void M1([NotNullWhenFalse] string? s) => throw null;
+                Diagnostic(ErrorCode.ERR_AttributeRequiresBoolReturn, "NotNullWhenFalse").WithArguments("NotNullWhenFalseAttribute").WithLocation(6, 22)
+                );
+
+            VerifyAnnotationsInSource(c, "C.M1", None);
+            VerifyAnnotationsInSource(c, "C.M2", None);
+            VerifyAnnotationsInSource(c, "C.M3", None);
+        }
+
+        [Fact]
+        public void NotNullWhenFalse_ReturningDynamic()
+        {
+            CSharpCompilation c = CreateCompilation(@"
+using System.Runtime.CompilerServices;
+public class C
+{
+    void Main(string? s)
+    {
+        if (string.IsNullOrWhiteSpace(s))
+        {
+            s.ToString(); // warn
+        }
+        else
+        {
+            s.ToString(); // ok
+        }
+    }
+    public dynamic MyIsNullOrEmpty([NotNullWhenFalse] string? s) => throw null;
+}
+" + NotNullWhenFalseAttributeDefinition, parseOptions: TestOptions.Regular8);
+
+            c.VerifyDiagnostics(
+                // (16,37): error CS8627: The NotNullWhenFalse attribute is only applicable on members that return a boolean type.
+                //     public dynamic MyIsNullOrEmpty([NotNullWhenFalse] string? s) => throw null;
+                Diagnostic(ErrorCode.ERR_AttributeRequiresBoolReturn, "NotNullWhenFalse").WithLocation(16, 37),
+                // (9,13): warning CS8602: Possible dereference of a null reference.
+                //             s.ToString(); // warn
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s").WithLocation(9, 13)
+                );
+
+            VerifyAnnotationsInSource(c, "C.MyIsNullOrEmpty", None);
+        }
+
+        [Fact]
+        public void NotNullWhenFalse_ReturningObject_FromMetadata()
+        {
+            string il = @"
```

Comment:
The intent is to validate the loading of misused in metadata. But that metadata can't be compiled because we produce an error for misused attribute in source.

---
In reply to: [187463607](https://github.com/dotnet/roslyn/pull/26656#discussion_r187463607) [](ancestors = 187463607)